Web Developer Puzzle-Task 4- Feat/mark as read

abhifoyou opened the pull request:
Implemented the Task 4 where user can mark a book as finished. Also implemented the related e2e test case.

![image](https://user-images.githubusercontent.com/25454641/94863342-4e02fc00-0408-11eb-82f1-988d00c011b8.png)

![image](https://user-images.githubusercontent.com/25454641/94863549-9c17ff80-0408-11eb-9fc3-2707a91a4aa0.png)

abhifoyou commented:
@jdpearce Request you to review.

mandarini commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -17,4 +17,55 @@ describe('When: I use the reading list feature', () => {
+    const toBeFinishedBooks = await $$(
+      '.reading-list-item [data-testing="mark-as-finished"]'
+    );
+    expect(toBeFinishedBooks.length > 0).toBeTruthy();
```
What if there was already a book in state that it was marked as read? You are supposing the state is clean. To catch this, maybe you should check the number of finished books, and then check if that number got augmented by 1.

abhifoyou replied:
The e2e does not fail here as i am explicitly adding a book in the previous spec in this test suite. So, when this spec runs but obvious there would be at least a book which is NOT finished/read.

mandarini commented on libs/api/books/src/lib/reading-list.controller.ts:
```diff
@@ -20,4 +20,10 @@ export class ReadingListController {
+  @Put('/reading-list/:id/finished')
+  async markAsFinished(@Param() params) {
+    const readingListItemArray = await this.readingList.markAsFinished(params.id);
+      return (readingListItemArray.length === 0) ? "No data found" : readingListItemArray[0];
```
There are a few extra spaces before the return statement. Also, can you please explain why you are following this approach here?

abhifoyou replied:
Ran Prettier, should be formatted now. 

The reason for this approach is in the next comment.

mandarini commented on libs/api/books/src/lib/reading-list.service.ts:
```diff
@@ -28,4 +28,18 @@ export class ReadingListService {
+          })
+    
+        });
+        return this.storage.read().filter(readingitem => readingitem.bookId === id);
```
Can you explain your approach here?

abhifoyou replied:
I could have used Promise<void> instead of Promise<ReadingListItem[]> and return the readingListItem array accordingly. Below is the code what I did before and somehow the view was not getting updated. May be I was doing something wrong.

`//reading-list.controller.ts
  
  @Put('/reading-list/:id/finished')
  async markAsFinished(@Param() params, @Body() item: ReadingListItem) {
  return await this.readingList.markAsFinished(params.id, item);
  }`

` //reading-list.service.ts
  
  async markAsFinished(id: string, updates: ReadingListItem): Promise<void> {
  this.storage.update(list => {
  return list.map(readingitem => {
  if(readingitem.bookId === id) {
  readingitem.finished = updates.finished;
  readingitem.finishedDate = updates.finishedDate
  }
  return readingitem;
  });
  });
  }`

And then I was going through some angular examples, it was advised to do a GET HTTP call to get all the data(Angular storage here) to be on the safer side and return once the promise is resolved. I was convinced to use the same and it worked fine too. 
  
However, if you could educate me on the difference on this 2 approaches and when to use what, that would be great.

mandarini commented on libs/api/books/src/lib/reading-list.service.ts:
```diff
@@ -28,4 +28,18 @@ export class ReadingListService {
+
+
+  async markAsFinished(id: string): Promise<ReadingListItem[]> {
+       this.storage.update(list => {
+          return list.map(readingitem => {
+            if (readingitem.bookId === id) {
+              readingitem = Object.assign({}, readingitem, { finished: true, finishedDate: new Date().toISOString() })
+            }
+            return readingitem;
+          })
+    
+        });
+        return this.storage.read().filter(readingitem => readingitem.bookId === id);
+      }
```
This is very indented. Did you forget to run prettier?

abhifoyou replied:
Ran Prettier, should be formatted now.

mandarini commented on libs/books/data-access/src/lib/+state/reading-list.effects.ts:
```diff
@@ -56,6 +57,25 @@ export class ReadingListEffects implements OnInitEffects {
+          );
+        },
+        undoAction: ({ item }) => {
+          return ReadingListActions.failedMarkAsFinished({item,error:"Couldn't Mark as Finished"})
```
It seems you forgot to run prettier

abhifoyou replied:
Ran Prettier, should be formatted now.

mandarini commented on libs/books/data-access/src/lib/+state/reading-list.effects.ts:
```diff
@@ -56,6 +57,25 @@ export class ReadingListEffects implements OnInitEffects {
+  markAsFinished$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(ReadingListActions.markAsFinished),
+      optimisticUpdate({
```
Why did you choose to use this method?

abhifoyou replied:
No specific reason :) . I am pretty much new to Angular. Read about the optimisticUpdate so wanted to give a try. Well, I know its advantages and disadvantages though. I could have done the same implementation like below. Not sure, if this is a ideal scenario to use optimisticUpdate though, would be great if you can throw some light. :)

![image](https://user-images.githubusercontent.com/25454641/95401925-798c5780-08dc-11eb-8c86-a1567d94ca15.png)

mandarini commented on libs/books/data-access/src/lib/+state/reading-list.effects.ts:
```diff
@@ -2,6 +2,7 @@ import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
 import { Actions, createEffect, ofType, OnInitEffects } from '@ngrx/effects';
 import { of } from 'rxjs';
+import { fetch, optimisticUpdate } from '@nrwl/angular';
```
fetch is not used, maybe it should not be imported

abhifoyou replied:
Removed.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -1,7 +1,11 @@
+      <img
+        class="reading-list-item--cover"
+        [src]="b.coverUrl"
+        alt="{{ b.title }}"
```
Since in other places you are using the "proper" angular template syntax for dynamic HTML attributes, why are you using this syntax here?

abhifoyou replied:
Fixed the same.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -1,18 +1,27 @@
+
+
 export class ReadingListComponent {
+  public ngUnsubscribe: Subject<boolean> = new Subject();
```
This is never used. Why did you add it here? How were you intending to use it?

abhifoyou replied:
It was not intended, removed the same.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -1,18 +1,27 @@
-import { Component } from '@angular/core';
+import { Component, OnInit, OnDestroy } from '@angular/core';
```
Why did you add these imports here, since they are not used?

abhifoyou replied:
Removed all the unused imports.

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -1,18 +1,27 @@
 import { Store } from '@ngrx/store';
-import { getReadingList, removeFromReadingList } from '@tmo/books/data-access';
+import { getReadingList, removeFromReadingList, markAsFinished} from '@tmo/books/data-access';
+import { takeUntil } from 'rxjs/operators';
```
Why did you add these import here, since it's not used?

abhifoyou replied:
Removed all the unused imports.

mandarini commented on libs/shared/styles/src/lib/variables.scss:
```diff
@@ -13,6 +13,7 @@ $gray85: #262626;
 $black: #000000;
 $white: #ffffff;
 $magenta:#e20074;
+$green:#228B22;
```
You should add a space between the semicolon and the hash sign.

abhifoyou replied:
Fixed.

mandarini reviewed (changes requested):
I think you should run prettier to format your files properly. Also, I would like to hear why you chose to go with some specific approaches! 

Also, the tests do not pass (e2e and unit tests fail)...

hanks for this PR :)

abhifoyou commented:
Fixed the unit tests too.

![image](https://user-images.githubusercontent.com/25454641/95402849-b9ecd500-08de-11eb-8511-a94bdffd4f61.png)

abhifoyou commented:
@mandarini : I have addressed all the review comments as part of this task. Please review and let me know if we are good.

Thanks for reviewing my PR. :)
